### PR TITLE
Adding support for Windows

### PR DESCRIPTION
**Pull request by nmaludy**
This is another PR to add support for Windows. Hopefully this is more concise than the others.

**TODO**
- [x] Need to test the `null` values for `facter::facts_d_mode` in the Windows data actually performs the Hiera lookup correctly.
- [x] Need to test the user/group of `NT AUTHORITY\SYSTEM` sets the proper user/group.
- [x] Fix unit tests (seems a bunch of them are broken)
- [x] Add spec tests for Windows OS

### CODE REVIEW

File: data/os/windows.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+facter::facts_d_dir: C:\ProgramData\PuppetLabs\facter\facts.d
+facter::facts_d_owner: NT AUTHORITY\SYSTEM
+facter::facts_d_group: NT AUTHORITY\SYSTEM
+facter::facts_d_mode: null
+facter::path_to_facter: C:\Program Files\Puppet Labs\Puppet\bin\facter.bat
```
Comment (ghoneycutt) [inline data/os/windows.yaml]: I believe these values with spaces will need to be quoted.

File: manifests/fact.pp [hunk 2]
```diff
@@ -4,27 +4,34 @@
 #
 define facter::fact (
   $value,
-  $fact      = $name,
-  $file      = 'facts.txt',
```
Comment (ghoneycutt) [inline manifests/fact.pp]: Would prefer if the default values stay here and then the hiera data for windows can override. That way you don't have to use `Optional` to set `undef` since they are not really optional. It is also a pattern I use elsewhere.

File: manifests/init.pp [hunk 3]
```diff
@@ -3,85 +3,43 @@
...
-  $facts_hash             = {},
-  $facts_hash_hiera_merge = false,
-  $facts_file             = 'facts.txt',
-  $facts_file_owner       = 'root',
-  $facts_file_group       = 'root',
-  $facts_file_mode        = '0644',
+  Boolean                    $manage_facts_d_dir     = true,
+  Boolean                    $purge_facts_d          = false,
+  Stdlib::Absolutepath       $facts_d_dir            = '/etc/facter/facts.d',
+  String                     $facts_d_owner          = 'root',
```
Comment (ghoneycutt) [inline manifests/init.pp]: owner/group use `String[1]`

File: manifests/init.pp [hunk 4]
```diff
@@ -3,85 +3,43 @@
...
-  $path_to_facter         = '/usr/bin/facter',
-  $path_to_facter_symlink = '/usr/local/bin/facter',
-  $ensure_facter_symlink  = false,
-  $facts_hash             = {},
-  $facts_hash_hiera_merge = false,
-  $facts_file             = 'facts.txt',
-  $facts_file_owner       = 'root',
-  $facts_file_group       = 'root',
-  $facts_file_mode        = '0644',
+  Boolean                    $manage_facts_d_dir     = true,
```
Comment (ghoneycutt) [inline manifests/init.pp]: just one space between things like

`DataType $variable = 'value',`

With the arrival of data types, keeping things in columns doesn't make sense anymore as the columns can get WAY WAY spaced out and when you add one it causes a bunch of code churn to respace things.

File: metadata.json [hunk 5]
```diff
@@ -10,10 +10,10 @@
   "issues_url": "https://github.com/ghoneycutt/puppet-module-facter/issues",
   "description": "Manage Facter and optionally ensure facts.d directory.",
   "requirements": [
-    {"name":"puppet","version_requirement":">= 3.0.0 < 7.0.0"}
+    {"name":"puppet","version_requirement":">= 4.7.0 < 7.0.0"}
```
Comment (ghoneycutt) [inline metadata.json]: This could be 5.8.. I certainly don't care about puppet 4 and don't want to bother testing against it.

**Comment by ghoneycutt** [comment]
@nmaludy this is awesome work, thank you very much!!

**Comment by nmaludy** [comment]
OK, i got Travis to the point where it's passing linting and running the spec tests. Most of the specs are failing currently. I'll fix those next.

File: spec/fixtures/hiera/hiera.yaml [hunk 6]
```diff
@@ -1,8 +1,10 @@
...
-:backends:
-  - yaml
-:yaml:
-  :datadir: 'spec/fixtures/hiera'
-:hierarchy:
-  - fqdn/%{fqdn}
-  - parameter_tests/%{parameter_tests}
+version: 5
+defaults:
+  datadir: data          # This path is relative to hiera.yaml's directory.
```
Comment (ghoneycutt) [inline spec/fixtures/hiera/hiera.yaml]: Wouldn't this keep it from using the data in fixtures which is the point of this file?

File: spec/fixtures/hiera/hiera.yaml [hunk 7]
```diff
@@ -1,8 +1,10 @@
 ---
-:backends:
-  - yaml
-:yaml:
-  :datadir: 'spec/fixtures/hiera'
-:hierarchy:
-  - fqdn/%{fqdn}
-  - parameter_tests/%{parameter_tests}
+version: 5
```
Comment (ghoneycutt) [inline spec/fixtures/hiera/hiera.yaml]: I thought that the spec tests still had to use the old version. Very cool if this now works!

File: spec/fixtures/hiera/hiera.yaml (hunk 6 above)
Comment (nmaludy) [reply spec/fixtures/hiera/hiera.yaml]: No, sorry the `spec` tests are configured to use this `spec/fixtures/hiera/hiera.yaml` and the `datadir` is relative to that file's directory, so it uses: `spec/fixtures/hiera/data`. Does that make sense? Sorry for the confusing wording.

File: spec/fixtures/hiera/hiera.yaml (hunk 7 above)
Comment (nmaludy) [reply spec/fixtures/hiera/hiera.yaml]: wooo! party on!

**Comment by nmaludy** [comment]
@ghoneycutt got tests written, the only one that is failing is https://github.com/ghoneycutt/puppet-module-facter/pull/67/files#diff-279972940b2611b2c7a177c5e1344902R133-R136

I think we do need the `Optional[]` around the `$file` and `$facts_dir` so we can fall back to using `$facter::facts_file` and `$facter::facts_d_dir`. Is that OK?
